File: src-tauri/src/services/gateway_api_service.rs
use std::time::Duration;

use reqwest::{Client, Method};
use serde_json::{json, Value};

use crate::models::error::{AppError, ErrorCode};
use crate::services::gateway_service;

const GATEWAY_HTTP_TIMEOUT_MS: u64 = 10_000;

fn normalize_path(path: &str) -> String {
    if path.starts_with('/') {
        path.to_string()
    } else {
        format!("/{path}")
    }
}

fn normalize_base(base: &str) -> String {
    base.trim_end_matches('/').to_string()
}

fn build_gateway_url(base: &str, path: &str) -> String {
    format!("{}{}", normalize_base(base), normalize_path(path))
}

fn build_http_client() -> Result<Client, AppError> {
    Client::builder()
        .timeout(Duration::from_millis(GATEWAY_HTTP_TIMEOUT_MS))
        .build()
        .map_err(|error| {
            AppError::new(
                ErrorCode::InternalError,
                "Failed to initialize HTTP client for Gateway proxy.",
                "Retry the request. If this persists, inspect desktop runtime logs.",
            )
            .with_details(json!({
                "sourceError": error.to_string(),
            }))
        })
}

async fn resolve_gateway_base_url() -> Result<String, AppError> {
    let status = gateway_service::get_gateway_status().await?;
    if !status.running {
        return Err(AppError::new(
            ErrorCode::GatewayNotRunning,
            "Gateway is not running, API proxy request cannot be completed.",
            "Start Gateway first, then retry the API operation.",
        )
        .with_details(json!({
            "address": status.address,
            "port": status.port,
            "state": status.state,
        })));
    }

    Ok(normalize_base(&status.address))
}

fn parse_response_value(raw: &str) -> Value {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return json!({});
    }

    serde_json::from_str::<Value>(trimmed).unwrap_or_else(|_| {
        json!({
            "raw": raw,
        })
    })
}

pub async fn request_gateway_json(
    method: Method,
    paths: &[&str],
    body: Option<Value>,
) -> Result<Value, AppError> {
    if paths.is_empty() {
        return Err(AppError::new(
            ErrorCode::InvalidInput,
            "Gateway request has no candidate paths.",
            "Provide at least one API path before sending a Gateway request.",
        ));
    }

    let base = resolve_gateway_base_url().await?;
    let client = build_http_client()?;
    let mut last_status: Option<u16> = None;
    let mut last_body: Option<String> = None;
    let mut last_url: Option<String> = None;
    let mut last_transport_error: Option<String> = None;

    for path in paths {
        let url = build_gateway_url(&base, path);
        let mut request = client.request(method.clone(), &url);
        if let Some(payload) = body.as_ref() {
            request = request.json(payload);
        }

        match request.send().await {
            Ok(response) => {
                let status = response.status();
                let text = response.text().await.unwrap_or_default();
                if status.is_success() {
                    return Ok(parse_response_value(&text));
                }

                last_status = Some(status.as_u16());
                last_body = Some(text);
                last_url = Some(url);
            }
            Err(error) => {
                last_transport_error = Some(error.to_string());
                last_url = Some(url);
            }
        }
    }

    let mut details = json!({
        "method": method.as_str(),
        "paths": paths,
        "lastStatus": last_status,
        "lastResponseBody": last_body,
        "lastUrl": last_url,
        "transportError": last_transport_error,
    });

    if let Some(map) = details.as_object_mut() {
        if map.get("lastStatus").is_some_and(Value::is_null) {
            map.remove("lastStatus");
        }
        if map.get("lastResponseBody").is_some_and(Value::is_null) {
            map.remove("lastResponseBody");
        }
        if map.get("lastUrl").is_some_and(Value::is_null) {
            map.remove("lastUrl");
        }
        if map.get("transportError").is_some_and(Value::is_null) {
            map.remove("transportError");
        }
    }

    Err(AppError::new(
        ErrorCode::NetworkFailed,
        "Gateway API request failed across all fallback paths.",
        "Check Gateway route availability and network diagnostics, then retry.",
    )
    .with_details(details))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalizes_paths_and_base_url() {
        assert_eq!(
            build_gateway_url("http://127.0.0.1:4317/", "api/channels"),
            "http://127.0.0.1:4317/api/channels"
        );
        assert_eq!(
            build_gateway_url("http://127.0.0.1:4317", "/api/channels"),
            "http://127.0.0.1:4317/api/channels"
        );
    }

    #[test]
    fn parses_empty_and_plain_text_response() {
        assert_eq!(parse_response_value("   "), json!({}));
        assert_eq!(
            parse_response_value("plain-text"),
            json!({ "raw": "plain-text" })
        );
    }
}

File: src/main/services/gateway-api-service.ts
import { AppError, ErrorCode } from "../models/error.js";
import { getGatewayStatus } from "./gateway-service.js";

const GATEWAY_HTTP_TIMEOUT_MS = 10_000;

function normalizeBase(base: string): string {
  return base.replace(/\/$/, "");
}

function normalizePath(p: string): string {
  return p.startsWith("/") ? p : `/${p}`;
}

function buildGatewayUrl(base: string, p: string): string {
  return `${normalizeBase(base)}${normalizePath(p)}`;
}

async function resolveGatewayBaseUrl(): Promise<string> {
  const status = await getGatewayStatus();
  if (!status.running) {
    throw new AppError(
      ErrorCode.GatewayNotRunning,
      "Gateway is not running, API proxy request cannot be completed.",
      "Start Gateway first, then retry the API operation.",
      { address: status.address, port: status.port, state: status.state },
    );
  }
  return normalizeBase(status.address);
}

function parseResponseValue(raw: string): unknown {
  const trimmed = raw.trim();
  if (!trimmed) return {};
  try {
    return JSON.parse(trimmed);
  } catch {
    return { raw };
  }
}

export async function requestGatewayJson(
  method: string,
  paths: string[],
  body?: unknown,
): Promise<unknown> {
  if (paths.length === 0) {
    throw new AppError(ErrorCode.InvalidInput, "Gateway request has no candidate paths.", "Provide at least one API path before sending a Gateway request.");
  }

  const base = await resolveGatewayBaseUrl();

  let lastStatus: number | undefined;
  let lastBody: string | undefined;
  let lastUrl: string | undefined;
  let lastTransportError: string | undefined;

  for (const p of paths) {
    const url = buildGatewayUrl(base, p);
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), GATEWAY_HTTP_TIMEOUT_MS);

    try {
      const headers: Record<string, string> = { Accept: "application/json" };
      const hasBody = body !== undefined && body !== null && method !== "GET" && method !== "DELETE";
      if (hasBody) headers["Content-Type"] = "application/json";

      const response = await fetch(url, {
        method,
        headers,
        body: hasBody ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });
      clearTimeout(timer);

      const text = await response.text().catch(() => "");
      if (response.ok) {
        return parseResponseValue(text);
      }
      lastStatus = response.status;
      lastBody = text;
      lastUrl = url;
    } catch (err: unknown) {
      clearTimeout(timer);
      lastTransportError = (err as Error)?.message ?? String(err);
      lastUrl = url;
    }
  }

  if (lastTransportError) {
    throw new AppError(
      ErrorCode.NetworkFailed,
      "Gateway API request failed due to a network error.",
      "Check whether Gateway is running and the network is accessible.",
      { url: lastUrl, transportError: lastTransportError },
    );
  }

  throw new AppError(
    ErrorCode.InternalError,
    `Gateway API request failed with HTTP ${lastStatus}.`,
    "Check Gateway logs and ensure the API endpoint is correct.",
    { url: lastUrl, status: lastStatus, body: lastBody?.slice(0, 500) },
  );
}
